### PR TITLE
Feature/controllers tests

### PR DESCRIPTION
**Pull request by volodymyrZotov**

### CODE REVIEW

File: controllers/deployment_controller_test.go [hunk 1]
```diff
@@ -19,99 +19,316 @@ import (
...
 					return false
 				}
 				return true
 			}, timeout, interval).Should(BeTrue())
-			Expect(createdSecret.Data).Should(Equal(expectedSecretData))
+			Expect(updatedSecret.Data).Should(Equal(item1.SecretData))
+		})
+
+		It("Should not delete secret created via deployment if it's used in another container", func() {
+			By("Create another POD with created secret")
```
Comment (edif2008) [inline controllers/deployment_controller_test.go:278]: Nit:
```suggestion
			By("Creating another POD with created secret")
```

**Review by edif2008 — COMMENTED** [review]
Code review: ✅ 
These tests look really solid. Thank you for migrating these test cases. Much appreciated.

I have a couple of notes:
- For the deployment controller tests, the Before section has `By` parts that show up even before providing the `Context` or the `It` part. This ends up with an odd ordering of the actions, which makes debugging a bit nasty.
- Some descriptions seem to focus on the end result rather than what actually happens in that test case. We can discuss more about these in a call.

File: controllers/onepassworditem_controller_test.go [hunk 2]
```diff
@@ -24,42 +24,36 @@ const (
 	iceCream  = "freezing blue 20%"
 )
 
-var _ = Describe("OnePasswordItem controller", func() {
+var _ = Describe("TestItem controller", func() {
```
[the remark this answers is not in the document]
Comment (volodymyrZotov) [reply controllers/onepassworditem_controller_test.go:27]: Looks like it was changed mistakenly. Definitely should revert it.

File: controllers/onepassworditem_controller_test.go [hunk 3]
```diff
@@ -240,5 +234,225 @@ var _ = Describe("OnePasswordItem controller", func() {
...
+			Expect(k8sClient.Create(ctx, toCreate)).Should(Succeed())
+
+			By("Reading K8s secret")
+			secret := &v1.Secret{}
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, key, secret)
+				return err == nil
+			}, timeout, interval).Should(BeTrue())
+			Expect(secret.Type).Should(Equal(v1.SecretTypeBasicAuth))
+		})
```
[the remark this answers is not in the document]
Comment (volodymyrZotov) [reply controllers/onepassworditem_controller_test.go:324]: This test doesn't make any sense as `Type` field is immutable. I was trying to change type from `SecretTypeBasicAuth` to `SecretTypeOpaque`.
Kubernetes throws an error if try to change it.
```Secret "test5" is invalid: type: Invalid value: "Opaque": field is immutable```.

[the remark this answers is not in the document]
Comment (volodymyrZotov) [reply controllers/onepassworditem_controller_test.go:324]: There is already a suite for that `It("Should throw an error if K8s Secret type is changed", func() {` line 333

File: controllers/onepassworditem_controller_test.go [hunk 4]
```diff
@@ -240,5 +234,225 @@ var _ = Describe("OnePasswordItem controller", func() {
...
+				if err != nil {
+					return false
+				}
+				return true
+			}, timeout, interval).Should(BeTrue())
+			Expect(secret.Type).Should(Equal(v1.SecretType(customType)))
+		})
+	})
+
+	Context("Failing part", func() {
```
[the remark this answers is not in the document]
Comment (volodymyrZotov) [reply controllers/onepassworditem_controller_test.go:363]: done

File: controllers/onepassworditem_controller_test.go [hunk 5]
```diff
@@ -240,5 +234,225 @@ var _ = Describe("OnePasswordItem controller", func() {
...
+			secret := &v1.Secret{}
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, key, secret)
+				if err != nil {
+					return false
+				}
+				return true
+			}, timeout, interval).Should(BeTrue())
+
+			By("Updating K8s secret type throw an error")
```
[the remark this answers is not in the document]
Comment (volodymyrZotov) [reply controllers/onepassworditem_controller_test.go:396]: changed

File: controllers/onepassworditem_controller_test.go [hunk 6]
```diff
@@ -240,5 +234,225 @@ var _ = Describe("OnePasswordItem controller", func() {
...
+				}
+
+				By("Creating a new TestItem")
+				Expect(k8sClient.Create(ctx, toCreate)).To(HaveOccurred())
+
+			})
+		})
+
+		When("TestItem name contains capital letters", func() {
+			It("An error occurred", func() {
```
[the remark this answers is not in the document]
Comment (volodymyrZotov) [reply controllers/onepassworditem_controller_test.go:434]: updated

File: controllers/deployment_controller_test.go [hunk 7]
```diff
@@ -123,15 +340,79 @@ var _ = Describe("Deployment controller", func() {
...
 			}, timeout, interval).Should(Succeed())
 
+			Eventually(func() error {
+				f := &v1.Secret{}
+				return k8sClient.Get(ctx, secretKey, f)
+			}, timeout, interval).Should(Succeed())
+		})
+
+		It("Should not delete secret created via deployment if it's used in another volume", func() {
+			By("Create another POD with created secret")
```
[the remark this answers is not in the document]
Comment (volodymyrZotov) [reply controllers/deployment_controller_test.go:350]: updated

File: controllers/deployment_controller_test.go (hunk 1 above)
Comment (volodymyrZotov) [reply controllers/deployment_controller_test.go:278]: updated

**Comment by volodymyrZotov** [comment]
@edif2008 `Context`, `Describe`, `When` is the same suite wrappers (they call them node containers in ginkgo world). And they exist for the same purpose - to organize the different aspects of code that we are testing hierarchically.

There is no restriction on how you should use them (for example it's ok to use `Context` inside `Describe`, or `By` inside `Context`, or `Describe` inside `Context`.)

The one general rule here is that `It` (subject node) should be inside `Context`/`Describe` (node containers).
[That article](https://onsi.github.io/ginkgo/#organizing-specs-with-container-nodes) might help to understand what ginkgo does under the hood.

**Review by edif2008 — APPROVED** [review]

**Pull request merged**